Add API to remove items from stateful.Bundle

jomuel opened the pull request:
Implement the `consumes()` function proposed by @DRMacIver for this purpose.

Closes #136

Zac-HD commented on hypothesis-python/src/hypothesis/stateful.py:
```diff
@@ -361,9 +384,15 @@ def rule(targets=(), target=None, **kwargs):
+
+    Any other kwargs will be turned into a strategy and values from that
+    will be provided.
+
```
nit: drop empty lines at end of docstring.

Zac-HD commented on hypothesis-python/tests/cover/test_stateful.py:
```diff
@@ -352,6 +353,43 @@ def execute_step(self, step):
+        assert value1 != value2
+
+
+def test_consumes():
```
Idiom: nstead of this function, we usually just assign `TestMachineWithConsumingRule = MachineWithConsumingRule.TestCase()` and let pytest discover the resulting class.

I'd also love to see a test or rule that uses collections of values, e.g. `lists(consumes(b1))`.

jomuel commented:
Thanks, that's really nice. :smile: I'm happy to contribute to an innovative project like Hypothesis.

I addressed your comments except for the one with the additional test. When adding a rule like that I realized that using `consumes` within a strategy like `lists` does not work at the moment, as it would require `consumes` to return a `SearchStrategy`. I will look into how to make it work later, it should not be too hard.

Zac-HD commented:
Perhaps the new I'll-call-it-a `BundleConsumer` class can inherit from `Bundle` to be a `SearchStrategy`, to simplify the drawing code?  I agree that this should be straightforward once you work out a nice way to do it.

And glad to hear you're enjoying contributing :grin:

jomuel commented:
Good idea @Zac-HD, I made `Consumer` a subclass of `Bundle` and it seems to work.

jomuel commented on hypothesis-python/src/hypothesis/stateful.py:
```diff
@@ -288,20 +290,39 @@ def do_draw(self, data):
-        return bundle[cu.integer_range(data, 0, len(bundle) - 1, center=len(bundle))]
+        position = cu.integer_range(data, 0, len(bundle) - 1, center=len(bundle))
+        if self.consume:
+            return bundle.pop(position)
```
I did not change this now as I didn't know when the sentinel values are supposed to be removed.

DRMacIver replied:
The idea is to never remove the sentinel values from the list but instead to leave them there and simply hide that fact when generating, but don't worry about it for now!

jomuel commented:
Thanks for your reviews, I made the changes to the docstrings you both suggested. Now that I'm familiar with the stateful module already, I'd like to work on issue #321 too. :slightly_smiling_face:
